openmsupply/application-manager-web-app
# 347 revalidation logic

Fixes #347 (again)

### Branched from 348


### Prelude

Original logic didn't quite work as I hoped. When form element is updated i thought it was safe to run validation via useGetFullApplicationStructure, but it's not (responses are not refetched yet).

Will update diagram sometime

So added a few timestamp state variables to the hook.

Essentially the processing part of the hook will only re run:

* When data was changed (keeping track of last fetch and last process timestamps)
* When validation is required and last update of an element is < last fetch (so if it's not quite yet fetched, will revalidate next time it's fetched)

### Changes

Sorry for the extra bits that were change, I thought they were necessary

#### useGetFullApplicationStructure

* Consolidated fullStructure to also return responses and lastValidationTimestamp (helped with reducing amount of time ApplicationPage is "called")
*
* Removed all of the loading statuses (as per above)
* Now has two parameters or revalidation `revalidateAfterTimestamp` and `shouldRevalidate`, revalidateAfterTimestamps says validation should be executed if data is fetched after it and shouldRevalidate is true
* Added small use effect to watch for data changes and update lastRefetchedTimestamp
* Now main use effect doesn't watch data but watched above and comparese with lastProcessedTimestamp 
* Did a little bit of reshuffle (sorry it's stylistic I know, but my outdated branch that was tested had it that way)

### ApplicationPage

* Added types (changed strict indicator soz)
* Added RevalidationState (all in one to reducing amount of time ApplicationPage is "called")
* Added use effect and revalidation method
* Uses `useFormElementUpdateTracker` to set correct props for useGetFullApplicationStructure -> 
```
const shouldRevalidate = isLastElementUpdateProcessed && revalidationState.shouldProcessValidation
const revalidateAfterTimestamp = shouldRevalidate ? elementUpdatedTimestamp : 0
```

Above means, revalidate when data is refetcehed after element was updated.

### ApplicationViewWrapper

* Now uses `useFormElementUpdateTracker`

### useFormElementUpdateTracker

* Simple context state management for three fields:

`elementUpdatedTimestamp`

`elementEnteredTimestamp`

and `isLastElementUpdateProcessed` (deduce from above)

It's basically a coupy of ApplicationState (and trimmed)

### Testing

Using front end 347-testing and back end 347-(front-end)-test

`347-(front-end)-test`

Changes organisation application slightly and extends api end point for checking uniquenss. Organisation application first three questions
* First one same as before (validates uniquenss of company name)
* Second one, just a text fields
* Third one has validation that concats both first and second and checks uniequness

`347-testing`

Just added an onClick to a button to console log state of element 3 at the time of revalidation finishing

The idea behind this is, when we jump focus from first straight to submit button click, we wait for first field to be save (after it was validated internally through ApplicationViewWrapper). Full validation is queued to happen in `useGetFullApplicationStructure` after update to that field. Check out console log.

### To Test

Check out and init both branches.

Create new org application (might be a bit slow)

Change url to applicationNew

Enter something in all three fields.

Enter 'Us' in second field

Enter 'Drug-R-' in second field and without pressing tab or clicking anywhere press submit

See console output under `Re validated field -> `, should see that valid is false.

Can delete a character from first field and press submit, it should be true now
